### PR TITLE
KAA-925 Factor out common manager API

### PR DESCRIPTION
**Pull request by rasendubi**
/cc @forGGe

### CODE REVIEW

File: client/client-multi/client-c/src/kaa/kaa_extension.c [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+ * Return the index of extension in the kaa_extensions.
+ */
+static size_t extension_id_to_idx(kaa_extension_id id)
+{
+    for (size_t i = 0; i < EXTENSION_COUNT; ++i) {
+        if (kaa_extensions[i]->id == id) {
+            return i;
+        }
+    }
+    return -1;
```
Comment (forGGe) [inline client/client-multi/client-c/src/kaa/kaa_extension.c]: Signed\unsiged implicit cast. ssize_t is more applicable here.

Comment (forGGe) [inline client/client-multi/client-c/src/kaa/kaa_extension.c]: Or just provide macro INVAL_EXTENSION (or something like that) 
with body of (size_t)-1

File: client/client-multi/client-c/src/kaa/kaa_extension.c [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    for (; i < EXTENSION_COUNT; ++i) {
+        result = kaa_extensions[i]->init(kaa_context, &kaa_extension_contexts[i]);
+        if (result != KAA_ERR_NONE) {
+            break;
+        }
+    }
+
+    if (result != KAA_ERR_NONE) {
+        while (i > 0) {
+            --i;
```
Comment (forGGe) [inline client/client-multi/client-c/src/kaa/kaa_extension.c:81]: Two above can be merged

File: client/client-multi/client-c/src/kaa/kaa_extension_private.h [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+ */
+#include <kaa_common.h>
+
+#include "kaa_private.h"
+
+// Extension is not really an extension. It doesn't have any context
+// and its implementation is embedded in platfrom_protocol.
+
+// They will be extern, thus prefixed.
+static struct kaa_extension kaa_extension_bootstrap = {
```
Comment (forGGe) [inline client/client-multi/client-c/src/kaa/kaa_extension_private.h:37]: To keep things simpler we can introduce macro that will initialize kaa_extension struct and use it  kaa_extension[] array directly

File: client/client-multi/client-c/src/kaa/kaa_extension_private.h [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+ * must only be used from that file. If you need extension instance,
+ * use interface defined in the kaa_extension.h. (That is the reason
+ * why the file doesn't have include guard.)
+ */
+#include <kaa_common.h>
+
+#include "kaa_private.h"
+
+// Extension is not really an extension. It doesn't have any context
+// and its implementation is embedded in platfrom_protocol.
```
Comment (forGGe) [inline client/client-multi/client-c/src/kaa/kaa_extension_private.h]: Emphasise here if it going to be changed in future

File: client/client-multi/client-c/src/kaa/kaa_extension.c (hunk 2 above)
Comment (rasendubi) [inline client/client-multi/client-c/src/kaa/kaa_extension.c:81]: They can't. `for (; i > 0; --i)` have different semantics

File: client/client-multi/client-c/src/kaa/kaa_platform_protocol.c [hunk 5]
```diff
@@ -351,7 +351,9 @@ kaa_error_t kaa_platform_protocol_serialize_client_sync(kaa_platform_protocol_t
                                                       , size_t *buffer_size)
 {
     KAA_RETURN_IF_NIL4(self, info, buffer, buffer_size, KAA_ERR_BADPARAM);
```
Comment (forGGe) [inline client/client-multi/client-c/src/kaa/kaa_platform_protocol.c]: I think this macro worth to purify as well.

File: client/client-multi/client-c/src/kaa/kaa_extension_private.h (hunk 4 above)
Comment (rasendubi) [inline client/client-multi/client-c/src/kaa/kaa_extension_private.h]: That can't be changed. It's really embedded in platform protocol.

File: client/client-multi/client-c/src/kaa/kaa_extension_private.h (hunk 3 above)
Comment (rasendubi) [inline client/client-multi/client-c/src/kaa/kaa_extension_private.h:37]: I'm going to move `kaa_extension_bootstrap` to the bootstrap module later.

**Comment by forGGe** [comment]
Reviewed

File: client/client-multi/client-c/src/kaa/kaa_extension.c (hunk 2 above)
Comment (forGGe) [inline client/client-multi/client-c/src/kaa/kaa_extension.c:81]: Got it

File: client/client-multi/client-c/src/kaa/kaa_extension_private.h (hunk 3 above)
Comment (forGGe) [inline client/client-multi/client-c/src/kaa/kaa_extension_private.h:37]: Got it

File: client/client-multi/client-c/src/kaa/kaa_extension_private.h (hunk 4 above)
Comment (forGGe) [inline client/client-multi/client-c/src/kaa/kaa_extension_private.h]: =(

Comment (rasendubi) [inline client/client-multi/client-c/src/kaa/kaa_extension_private.h]: Wow. Just found out I've missed "meta data". Meta data extension is not a real extension.

**Comment by forGGe** [comment]
Looks good

**Pull request merged**